File: RiskAndroid/src/main/java/cc/game/android/risk/DiceView.java
package cc.game.android.risk;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.View;

import cc.lib.android.DroidGraphics;
import cc.lib.android.DroidView;
import cc.lib.game.AGraphics;
import cc.lib.game.GColor;
import cc.lib.game.Justify;
import cc.lib.game.Utils;
import cc.lib.utils.Lock;

/**
 * Created by Chris Caron on 8/20/21.
 */
public class DiceView extends DroidView implements Runnable, View.OnClickListener {

    int [] dieNums = { 6, 8, 9, 10, 12, 15, 20 };

    int finalNum=6;
    int dieNum = 0;
    int maxDieNums = 0;
    long delay = 0;
    boolean rolling = false;
    GColor dieColor = GColor.WHITE;
    GColor pipColor = GColor.BLACK;
    Lock lock = null;

    public DiceView(Context context, AttributeSet attrs) {
        super(context, attrs);
        TypedArray arr = context.obtainStyledAttributes(attrs, R.styleable.DiceView);
        dieColor = GColor.fromARGB(arr.getColor(R.styleable.DiceView_dieColor, dieColor.toARGB()));
        pipColor = GColor.fromARGB(arr.getColor(R.styleable.DiceView_pipColor, pipColor.toARGB()));
        arr.recycle();
    }

    @Override
    protected void onPaint(DroidGraphics g) {
        g.setLineWidth(5);
        if (rolling)
            drawDie(g, getHeight(), dieNum);
        else {
            drawDie(g, getHeight(), finalNum);
            g.setColor(GColor.CYAN);
            g.drawRoundedRect(0, 0, getWidth(), getHeight(), getWidth()/4);
        }
    }

    public void setLock(Lock lock) {
        this.lock = lock;
    }

    void setColors(GColor dieColor, GColor pipColor) {
        this.dieColor = dieColor;
        this.pipColor = pipColor;
        invalidate();
    }

    public synchronized void rollDice(int resultDie, Lock lock) {
        delay = 10;
        rolling = true;
        finalNum = resultDie;
        this.lock = lock;
        lock.acquire();
        run();
    }

    @Override
    public void onClick(View v) {
        delay = 10;
        rolling = true;
        run();
    }

    @Override
    public synchronized void run() {
        if (rolling) {
            dieNum = Utils.rand() % dieNums[maxDieNums] + 1;
            if (delay < 500) {
                delay += 20;
                postDelayed(this, delay);
            } else {
                rolling = false;
                if (lock != null) {
                    lock.release();
                }
            }
        }
        invalidate();
    }

    public void addPips() {
        if (maxDieNums < dieNums.length-1) {
            maxDieNums++;
            dieNum = dieNums[maxDieNums];
            invalidate();
        }
    }

    public void removePips() {
        if (maxDieNums > 0) {
            maxDieNums --;
            dieNum = dieNums[maxDieNums];
            invalidate();
        }
    }

    public void drawDie(AGraphics g, float dim, int pips) {
        g.setColor(dieColor);
        float arc = dim/4;
        g.drawFilledRoundedRect(0, 0, dim, dim, arc);
        g.setColor(pipColor);

        float dd2 = dim/2;
        float dd4 = dim/4;
        float dd34 = (dim*3)/4;
        float dotSize = dim/8;
        float oldDotSize = g.setPointSize(dotSize);
        if (dieNums[maxDieNums] > 9) {
            g.setTextHeight(dim/2);
            g.drawJustifiedString(dd2, dd2, Justify.CENTER, Justify.CENTER, String.valueOf(pips));
            return;
        }
        g.begin();
        switch (pips) {
            case 1:
                g.vertex(dd2, dd2);
                break;
            case 2:
                g.vertex(dd4, dd4);
                g.vertex(dd34, dd34);
                break;
            case 3:
                g.vertex(dd4, dd4);
                g.vertex(dd2, dd2);
                g.vertex(dd34, dd34);
                break;
            case 4:
                g.vertex(dd4, dd4);
                g.vertex(dd34, dd34);
                g.vertex(dd4, dd34);
                g.vertex(dd34, dd4);
                break;
            case 5:
                g.vertex(dd4, dd4);
                g.vertex(dd34, dd34);
                g.vertex(dd4, dd34);
                g.vertex(dd34, dd4);
                g.vertex(dd2, dd2);
                break;
            case 6:
                g.vertex(dd4, dd4);
                g.vertex(dd34, dd34);
                g.vertex(dd4, dd34);
                g.vertex(dd34, dd4);
                g.vertex(dd4, dd2);
                g.vertex(dd34, dd2);
                break;
            case 7:
                g.vertex(dd2, dd2);
                g.vertex(dd4, dd4);
                g.vertex(dd34, dd34);
                g.vertex(dd4, dd34);
                g.vertex(dd34, dd4);
                g.vertex(dd4, dd2);
                g.vertex(dd34, dd2);
                break;
            case 8:
                g.vertex(dd2, dd4);
                g.vertex(dd2, dd34);
                g.vertex(dd4, dd4);
                g.vertex(dd34, dd34);
                g.vertex(dd4, dd34);
                g.vertex(dd34, dd4);
                g.vertex(dd4, dd2);
                g.vertex(dd34, dd2);
                break;
            case 9:
                g.vertex(dd2, dd4);
                g.vertex(dd2, dd2);
                g.vertex(dd2, dd34);
                g.vertex(dd4, dd4);
                g.vertex(dd34, dd34);
                g.vertex(dd4, dd34);
                g.vertex(dd34, dd4);
                g.vertex(dd4, dd2);
                g.vertex(dd34, dd2);
                break;
            default:
                g.drawJustifiedString(dd2, dd2, Justify.CENTER, Justify.CENTER, String.valueOf(pips));
                return;
        }
        g.drawPoints();
        g.setPointSize(oldDotSize);
    }
}

File: RiskAndroid/src/main/java/cc/game/android/risk/DiceView.kt
package cc.game.android.risk

import android.content.Context
import android.util.AttributeSet
import android.view.View
import cc.lib.android.DroidGraphics
import cc.lib.android.DroidView
import cc.lib.game.AGraphics
import cc.lib.game.GColor
import cc.lib.game.Justify
import cc.lib.game.Utils
import cc.lib.utils.Lock

/**
 * Created by Chris Caron on 8/20/21.
 */
class DiceView(context: Context, attrs: AttributeSet?) : DroidView(context, attrs), Runnable, View.OnClickListener {
	var dieNums = intArrayOf(6, 8, 9, 10, 12, 15, 20)
	var finalNum = 6
	var dieNum = 0
	var maxDieNums = 0
	var delay: Long = 0
	var rolling = false
	var dieColor = GColor.WHITE
	var pipColor = GColor.BLACK
	lateinit var lock: Lock
	override fun onPaint(g: DroidGraphics) {
		g.setLineWidth(5f)
		if (rolling) drawDie(g, height.toFloat(), dieNum) else {
			drawDie(g, height.toFloat(), finalNum)
			g.color = GColor.CYAN
			g.drawRoundedRect(0f, 0f, width.toFloat(), height.toFloat(), (width / 4).toFloat())
		}
	}

	fun setColors(dieColor: GColor, pipColor: GColor) {
		this.dieColor = dieColor
		this.pipColor = pipColor
		postInvalidate()
	}

	@Synchronized
	fun rollDice(resultDie: Int, lock: Lock) {
		delay = 10
		rolling = true
		finalNum = resultDie
		this.lock = lock
		lock.acquire()
		run()
	}

	override fun onClick(v: View) {
		delay = 10
		rolling = true
		run()
	}

	@Synchronized
	override fun run() {
		if (rolling) {
			dieNum = Utils.rand() % dieNums[maxDieNums] + 1
			if (delay < 400) {
				delay += 25
				postDelayed(this, delay)
			} else {
				rolling = false
				lock.release()
			}
		}
		postInvalidate()
	}

	fun addPips() {
		if (maxDieNums < dieNums.size - 1) {
			maxDieNums++
			dieNum = dieNums[maxDieNums]
			postInvalidate()
		}
	}

	fun removePips() {
		if (maxDieNums > 0) {
			maxDieNums--
			dieNum = dieNums[maxDieNums]
			postInvalidate()
		}
	}

	fun drawDie(g: AGraphics, dim: Float, pips: Int) {
		g.color = dieColor
		val arc = dim / 4
		g.drawFilledRoundedRect(0f, 0f, dim, dim, arc)
		g.color = pipColor
		val dd2 = dim / 2
		val dd4 = dim / 4
		val dd34 = dim * 3 / 4
		val dotSize = dim / 8
		val oldDotSize = g.setPointSize(dotSize)
		if (dieNums[maxDieNums] > 9) {
			g.textHeight = dim / 2
			g.drawJustifiedString(dd2, dd2, Justify.CENTER, Justify.CENTER, pips.toString())
			return
		}
		g.begin()
		when (pips) {
			1 -> g.vertex(dd2, dd2)
			2 -> {
				g.vertex(dd4, dd4)
				g.vertex(dd34, dd34)
			}
			3 -> {
				g.vertex(dd4, dd4)
				g.vertex(dd2, dd2)
				g.vertex(dd34, dd34)
			}
			4 -> {
				g.vertex(dd4, dd4)
				g.vertex(dd34, dd34)
				g.vertex(dd4, dd34)
				g.vertex(dd34, dd4)
			}
			5 -> {
				g.vertex(dd4, dd4)
				g.vertex(dd34, dd34)
				g.vertex(dd4, dd34)
				g.vertex(dd34, dd4)
				g.vertex(dd2, dd2)
			}
			6 -> {
				g.vertex(dd4, dd4)
				g.vertex(dd34, dd34)
				g.vertex(dd4, dd34)
				g.vertex(dd34, dd4)
				g.vertex(dd4, dd2)
				g.vertex(dd34, dd2)
			}
			7 -> {
				g.vertex(dd2, dd2)
				g.vertex(dd4, dd4)
				g.vertex(dd34, dd34)
				g.vertex(dd4, dd34)
				g.vertex(dd34, dd4)
				g.vertex(dd4, dd2)
				g.vertex(dd34, dd2)
			}
			8 -> {
				g.vertex(dd2, dd4)
				g.vertex(dd2, dd34)
				g.vertex(dd4, dd4)
				g.vertex(dd34, dd34)
				g.vertex(dd4, dd34)
				g.vertex(dd34, dd4)
				g.vertex(dd4, dd2)
				g.vertex(dd34, dd2)
			}
			9 -> {
				g.vertex(dd2, dd4)
				g.vertex(dd2, dd2)
				g.vertex(dd2, dd34)
				g.vertex(dd4, dd4)
				g.vertex(dd34, dd34)
				g.vertex(dd4, dd34)
				g.vertex(dd34, dd4)
				g.vertex(dd4, dd2)
				g.vertex(dd34, dd2)
			}
			else -> {
				g.drawJustifiedString(dd2, dd2, Justify.CENTER, Justify.CENTER, pips.toString())
				return
			}
		}
		g.drawPoints()
		g.setPointSize(oldDotSize)
	}

	init {
		val arr = context.obtainStyledAttributes(attrs, R.styleable.DiceView)
		dieColor = GColor.fromARGB(arr.getColor(R.styleable.DiceView_dieColor, dieColor.toARGB()))
		pipColor = GColor.fromARGB(arr.getColor(R.styleable.DiceView_pipColor, pipColor.toARGB()))
		arr.recycle()
	}
}